experimental/InvariantTheory/invariant_rings.jl

```diff
@@ -290,57 +248,34 @@ julia> reynolds_operator(IR, f)
 0
 ```
 """
-reynolds_operator(IR::InvRing{FldT, GrpT, T}, f::T) where {FldT, GrpT, T <: MPolyElem} = reynolds_operator_via_oscar(IR, f)
-
-function reynolds_operator(IR::InvRing, f::MPolyElem)
-  @assert parent(f) === polynomial_ring(IR).R
-  return reynolds_operator(IR, polynomial_ring(IR)(f))
-end
-
-function basis_via_reynolds(IR::InvRing, d::Int)
-  @assert d >= 0 "Degree must be non-negative"
+function reynolds_operator(IR::InvRing{FldT, GrpT, T}, f::T) where {FldT, GrpT, T <: MPolyElem}
   @assert !ismodular(IR)
-  R = polynomial_ring(IR)
-  if d == 0
-    return elem_type(R)[ one(R) ]
-  end
+  @assert parent(f) === polynomial_ring(IR)
 
-  rey = reynolds_via_singular(IR)
-  basisSing = Singular.LibFinvar.invariant_basis_reynolds(rey, d)
-  res = Vector{elem_type(R)}()
-  # [ 0 ] is not a basis, let's return [ ]
-  if length(gens(basisSing)) == 1 && iszero(gens(basisSing)[1])
-    return res
-  end
-  for f in gens(basisSing)
-    push!(res, R(f))
+  if !isdefined(IR, :reynolds_operator)
+    _prepare_reynolds_operator(IR)
   end
-  return res
+  return IR.reynolds_operator(f)
 end
 
-function basis_via_linear_algebra(IR::InvRing, d::Int)
-  @assert d >= 0 "Degree must be non-negative"
-  R = polynomial_ring(IR)
-  if d == 0
-    return elem_type(R)[ one(R) ]
-  end
-
-  basisSing = Singular.LibFinvar.invariant_basis(d, _action_singular(IR)...)
-  res = Vector{elem_type(R)}()
-  # [ 0 ] is not a basis, let's return [ ]
-  if length(gens(basisSing)) == 1 && iszero(gens(basisSing)[1])
-    return res
-  end
-  for f in gens(basisSing)
-    push!(res, R(f))
-  end
-  return res
+function reynolds_operator(IR::InvRing, f::MPolyElem)
+  @assert parent(f) === polynomial_ring(IR).R
+  return reynolds_operator(IR, polynomial_ring(IR)(f))
 end
 
+################################################################################
+#
+#  Bases
+#
+################################################################################
+
 @doc Markdown.doc"""
-     basis(IR::InvRing, d::Int)
+     basis(IR::InvRing, d::Int, algo::Symbol = :default)
 
 Given an invariant ring `IR` and an integer `d`, return a basis for the invariants in degree `d`.
+The used algorithm can be specified using the optional argument `algo`. Possible values are `:reynolds` which uses the reynolds operator to construct the basis (only available in the non-modular case) and `:linear_algebra` which uses plain linear algebra. With the default value `:default` the heuristically best algorithm is selected.
```

Comment:
Please wrap long lines, so that they can be read in the GitHub web interface. I don't care whether it is after 80 or 90 or 100 chars... but it shouldn't bee much more per line.